File: runtime/native_uid/identity.cpp
#include "identity.h"
#include <algorithm>
#include <cmath>
#include <set>
#include <stdexcept>

namespace aii::uid {
namespace {
void require(bool ok,const char* why) { if(!ok) throw std::invalid_argument(why); }
bool sha(const std::string& s) {
  return s.size()==64 && s.find_first_not_of("0123456789abcdef")==std::string::npos;
}
double norm(const Vector& v) {
  double sum=0;
  for(const auto x:v) { require(std::isfinite(x),"nonfinite UID embedding"); sum+=x*x; }
  require(std::isfinite(sum) && sum>=1e-24,"invalid UID embedding norm");
  return std::sqrt(sum);
}
Vector unit(Vector v) { const auto n=norm(v); for(auto& x:v)x/=n; return v; }
bool label(const std::string& s) {
  try {
    const auto points=unicode_scalars(s);
    return !points.empty()&&points.size()<=128&&std::all_of(points.begin(),points.end(),[](uint32_t cp){return cp>=32;});
  }catch(const std::invalid_argument&){return false;}
}
void policy(const Policy& p) {
  require(sha(p.embedding_binding)&&sha(p.calibration_sha256)&&sha(p.fingerprint),"UID policy binding missing");
  require(std::isfinite(p.threshold)&&p.threshold>=-1&&p.threshold<=1,"invalid UID threshold");
  require(std::isfinite(p.minimum_margin)&&p.minimum_margin>=0&&p.minimum_margin<=2,"invalid UID margin");
  require(p.minimum_enrollment_samples>=1&&p.minimum_enrollment_samples<=8,"invalid enrollment minimum");
}
Vector centroid(const Speaker& s) {
  Vector v{};
  for(const auto& sample:s.samples) for(size_t i=0;i<v.size();++i)v[i]+=sample.embedding[i];
  // Match the established mean-then-normalize decision path, including its
  // double precision; samples themselves are never normalized on read.
  for(auto& x:v)x/=s.samples.size();
  return unit(v);
}
}
std::vector<uint32_t> unicode_scalars(const std::string& s) {
  std::vector<uint32_t> points;
  for(size_t i=0;i<s.size();) {
    const auto b=static_cast<unsigned char>(s[i++]);
    uint32_t cp=b; unsigned tail=0; uint32_t low=0;
    if(b>=0xc2 && b<=0xdf) {cp=b&31;tail=1;low=0x80;}
    else if(b>=0xe0 && b<=0xef) {cp=b&15;tail=2;low=0x800;}
    else if(b>=0xf0 && b<=0xf4) {cp=b&7;tail=3;low=0x10000;}
    else if(b>=0x80)throw std::invalid_argument("invalid UTF-8");
    for(unsigned j=0;j<tail;++j) {
      if(i==s.size())throw std::invalid_argument("truncated UTF-8");
      const auto c=static_cast<unsigned char>(s[i++]);
      if((c&0xc0)!=0x80)throw std::invalid_argument("invalid UTF-8 continuation");
      cp=(cp<<6)|(c&63);
    }
    if(cp<low || cp>0x10ffff || (cp>=0xd800 && cp<=0xdfff))throw std::invalid_argument("invalid Unicode scalar");
    points.push_back(cp);
  }
  return points;
}
void validate(const Snapshot& s,const Policy& expected) {
  policy(expected);policy(s.policy);
  const auto& p=s.policy;
  require(p.embedding_binding==expected.embedding_binding&&p.calibration_sha256==expected.calibration_sha256&&
          p.fingerprint==expected.fingerprint&&p.threshold==expected.threshold&&p.minimum_margin==expected.minimum_margin&&
          p.minimum_enrollment_samples==expected.minimum_enrollment_samples,"enrollment policy/model binding differs");
  require(s.revision<=INT64_MAX && s.speakers.size()<=256,"invalid enrollment revision/count");
  require(s.speakers.empty()||s.revision>0,"enrollment requires positive revision");
  std::set<std::string> audio;
  std::string previous;
  for(const auto& speaker:s.speakers) {
    const auto& id=speaker.id;
    const std::string alnum="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    require(!id.empty()&&id.size()<=96&&alnum.find(id[0])!=std::string::npos&&
            id.find_first_not_of(alnum+"_.-")==std::string::npos&&id>previous,"invalid/unsorted speaker ID");
    previous=id;
    require(label(speaker.label),"invalid enrollment label");
    require(!speaker.samples.empty()&&speaker.samples.size()<=8,"invalid enrollment sample count");
    std::string last;
    for(const auto& sample:speaker.samples) {
      require(sha(sample.audio_sha256)&&sample.audio_sha256>last&&audio.insert(sample.audio_sha256).second,
              "invalid/duplicate/unsorted enrollment audio");
      last=sample.audio_sha256;
      require(std::abs(norm(sample.embedding)-1)<=1e-6,"invalid enrollment unit vector");
    }
    (void)centroid(speaker);
  }
}
Decision identify(const Snapshot& s,const Policy& expected,const Vector& embedding,const std::string& binding) {
  validate(s,expected);
  require(binding==expected.embedding_binding,"embedding model/frontend binding differs");
  const auto query=unit(embedding);
  Decision result{"unknown","","","no_enrollments",std::nullopt,std::nullopt,s.revision,expected.fingerprint};
  if(s.speakers.empty())return result;
  std::vector<std::pair<double,size_t>> scores;
  for(size_t j=0;j<s.speakers.size();++j) {
    const auto center=centroid(s.speakers[j]);double dot=0;
    for(size_t i=0;i<center.size();++i)dot+=query[i]*center[i];
    scores.emplace_back(std::clamp(dot,-1.,1.),j);
  }
  std::sort(scores.begin(),scores.end(),[](const auto& a,const auto& b){
    return a.first==b.first?a.second<b.second:a.first>b.first;
  });
  const auto& winner=s.speakers[scores[0].second];result.score=scores[0].first;
  if(scores.size()>1)result.margin=scores[0].first-scores[1].first;
  if(winner.samples.size()<expected.minimum_enrollment_samples)result.reason="insufficient_enrollment";
  else if(*result.score<expected.threshold)result.reason="below_acceptance_threshold";
  else if(result.margin&&(*result.margin<=0||*result.margin<expected.minimum_margin)) {
    result.outcome="ambiguous";result.reason="insufficient_separation";
  } else {
    result.outcome="known";result.reason="accepted";result.speaker_id=winner.id;result.label=winner.label;
  }
  return result;
}
}

File: runtime/speaker_identity/identity.py
"""Bounded recognition decisions and transactional enrollment, without inference."""

from __future__ import annotations

import hashlib
import json
import math
import os
import re
import sqlite3
from dataclasses import asdict, dataclass
from pathlib import Path

import numpy as np

DIMENSIONS = 256
MAX_SPEAKERS = 256
MAX_SAMPLES = 8


class SpeakerIdentityError(ValueError):
    pass


def canonical(value) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), allow_nan=False)


def digest(value) -> str:
    return hashlib.sha256(canonical(value).encode()).hexdigest()


def sha_string(value: str) -> str:
    if not isinstance(value, str) or not re.fullmatch(r"[0-9a-f]{64}", value):
        raise SpeakerIdentityError("expected a lowercase SHA-256 identity")
    return value


def unit(value) -> np.ndarray:
    array = np.asarray(value, dtype=np.float64)
    if array.shape != (DIMENSIONS,) or not np.isfinite(array).all():
        raise SpeakerIdentityError("embedding must contain 256 finite values")
    norm = float(np.linalg.norm(array))
    if not math.isfinite(norm) or norm < 1e-12:
        raise SpeakerIdentityError("embedding has invalid norm")
    return array / norm


@dataclass(frozen=True)
class Policy:
    embedding_binding: str
    threshold: float
    minimum_margin: float
    calibration_sha256: str
    minimum_enrollment_samples: int = 1

    def __post_init__(self):
        sha_string(self.embedding_binding)
        sha_string(self.calibration_sha256)
        if (
            type(self.minimum_enrollment_samples) is not int
            or not 1 <= self.minimum_enrollment_samples <= MAX_SAMPLES
        ):
            raise SpeakerIdentityError("invalid minimum enrollment sample count")
        for name, low, high in (("threshold", -1, 1), ("minimum_margin", 0, 2)):
            value = getattr(self, name)
            if (
                isinstance(value, bool)
                or not math.isfinite(value)
                or not low <= value <= high
            ):
                raise SpeakerIdentityError(f"invalid {name}")

    @property
    def fingerprint(self):
        return digest(asdict(self))


@dataclass(frozen=True)
class Decision:
    outcome: str
    speaker_id: str | None
    label: str | None
    score: float | None
    margin: float | None
    reason: str
    enrollment_revision: int
    policy_sha256: str


def valid_id(value: str) -> str:
    if not isinstance(value, str) or not re.fullmatch(
        r"[A-Za-z0-9][A-Za-z0-9_.-]{0,95}", value
    ):
        raise SpeakerIdentityError(
            "speaker id must be 1..96 simple identifier characters"
        )
    return value


class IdentityStore:
    """One SQLite authority, explicit creation, exact policy binding, no audio stored.

    Each instance belongs to one thread. Independent instances/processes serialize
    enrollment transactions through SQLite. Inference runs before the transaction;
    identify reads one consistent enrollment snapshot and reports its revision.
    """

    @staticmethod
    def _initialize(connection, policy):
        with connection:
            connection.execute("BEGIN IMMEDIATE")
            connection.execute(
                "CREATE TABLE authority (singleton INTEGER PRIMARY KEY CHECK(singleton=1), policy TEXT NOT NULL, revision INTEGER NOT NULL CHECK(revision>=0))"
            )
            connection.execute(
                "CREATE TABLE speakers (id TEXT PRIMARY KEY, label TEXT NOT NULL)"
            )
            connection.execute(
                "CREATE TABLE samples (speaker_id TEXT NOT NULL REFERENCES speakers(id) ON DELETE CASCADE, audio_sha TEXT NOT NULL UNIQUE, embedding BLOB NOT NULL CHECK(length(embedding)=2048), PRIMARY KEY(speaker_id,audio_sha))"
            )
            connection.execute(
                "INSERT INTO authority VALUES (1,?,0)", (canonical(asdict(policy)),)
            )

    @classmethod
    def memory(cls, policy: Policy):
        """Explicit, thread-confined working copy; grants no persistence authority.

        Reuses the exact enrollment transactions and decision path. No temporary
        file, database deserialization, schema from a caller or model load.
        """
        self = cls.__new__(cls)
        self.path, self.policy = None, policy
        self.db = sqlite3.connect(":memory:")
        try:
            self.db.execute("PRAGMA foreign_keys=ON")
            self.db.execute("PRAGMA temp_store=MEMORY")
            self.db.execute("PRAGMA trusted_schema=OFF")
            self.db.execute("PRAGMA secure_delete=ON")
            cls._initialize(self.db, policy)
        except Exception:
            self.db.close()
            raise
        return self

    @classmethod
    def create(cls, path: Path, policy: Policy):
        path = Path(path)
        path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        # Do not silently replace or initialize an existing, damaged, or wrong DB.
        fd = os.open(path, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
        os.close(fd)
        connection = sqlite3.connect(path)
        try:
            connection.execute("PRAGMA synchronous=FULL")
            cls._initialize(connection, policy)
        finally:
            connection.close()
        return cls(path, policy)

    def __init__(self, path: Path, policy: Policy):
        self.path, self.policy = Path(path), policy
        self.db = sqlite3.connect(
            self.path.resolve().as_uri() + "?mode=rw", uri=True, timeout=2
        )
        try:
            self.db.execute("PRAGMA foreign_keys=ON")
            self.db.execute("PRAGMA synchronous=FULL")
            self.db.execute("PRAGMA secure_delete=ON")
            self._revision()
        except Exception:
            self.db.close()
            raise

    def close(self):
        self.db.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def _revision(self):
        rows = self.db.execute(
            "SELECT policy,revision FROM authority WHERE singleton=1"
        ).fetchall()
        if len(rows) != 1 or rows[0][0] != canonical(asdict(self.policy)):
            raise SpeakerIdentityError(
                "enrollment policy/model binding differs or is missing"
            )
        revision = rows[0][1]
        if type(revision) is not int or revision < 0:
            raise SpeakerIdentityError("invalid enrollment revision")
        return revision

    def _advance(self):
        self.db.execute("UPDATE authority SET revision=revision+1 WHERE singleton=1")
        return self._revision()

    def enroll(
        self,
        speaker_id: str,
        label: str,
        embedding,
        *,
        audio_sha256: str,
        embedding_binding: str,
    ):
        return self.enroll_many(
            speaker_id,
            label,
            [(audio_sha256, embedding)],
            embedding_binding=embedding_binding,
        )

    def enroll_many(self, speaker_id, label, recordings, *, embedding_binding):
        """Commit an explicitly selected recording set in one transaction.

        Inference is already complete. This reuses the single-sample authority
        and advances its revision once; no partial batch becomes visible.
        """
        valid_id(speaker_id)
        if (
            not isinstance(label, str)
            or not 1 <= len(label) <= 128
            or any(ord(c) < 32 for c in label)
        ):
            raise SpeakerIdentityError("label must be 1..128 printable characters")
        if embedding_binding != self.policy.embedding_binding:
            raise SpeakerIdentityError("embedding model/frontend binding differs")
        if (
            not isinstance(recordings, (list, tuple))
            or not 1 <= len(recordings) <= MAX_SAMPLES
        ):
            raise SpeakerIdentityError("bounded enrollment recordings required")
        prepared = [
            (sha_string(audio_sha), unit(vector)) for audio_sha, vector in recordings
        ]
        if len({audio_sha for audio_sha, _ in prepared}) != len(prepared):
            raise SpeakerIdentityError("duplicate evidence refused within enrollment")
        with self.db:
            self.db.execute("BEGIN IMMEDIATE")
            self._revision()
            if any(
                self.db.execute(
                    "SELECT 1 FROM samples WHERE audio_sha=?", (audio_sha,)
                ).fetchone()
                for audio_sha, _ in prepared
            ):
                raise SpeakerIdentityError(
                    "this audio is already enrolled; duplicate evidence refused"
                )
            existing = self.db.execute(
                "SELECT label FROM speakers WHERE id=?", (speaker_id,)
            ).fetchone()
            if existing and existing[0] != label:
                raise SpeakerIdentityError(
                    "existing speaker label differs; no silent identity replacement"
                )
            count = self.db.execute(
                "SELECT COUNT(*) FROM samples WHERE speaker_id=?", (speaker_id,)
            ).fetchone()[0]
            if count + len(prepared) > MAX_SAMPLES:
                raise SpeakerIdentityError("speaker enrollment sample limit reached")
            # A successful enrollment must leave a usable template, not plant a
            # zero/nonfinite centroid that breaks every later recognition call.
            prior = self.db.execute(
                "SELECT embedding FROM samples WHERE speaker_id=?", (speaker_id,)
            ).fetchall()
            unit(
                np.sum(
                    [unit(np.frombuffer(r[0], dtype="<f8")) for r in prior]
                    + [vector for _, vector in prepared],
                    axis=0,
                )
            )
            if not existing:
                if (
                    self.db.execute("SELECT COUNT(*) FROM speakers").fetchone()[0]
                    >= MAX_SPEAKERS
                ):
                    raise SpeakerIdentityError("speaker limit reached")
                self.db.execute(
                    "INSERT INTO speakers VALUES (?,?)", (speaker_id, label)
                )
            self.db.executemany(
                "INSERT INTO samples VALUES (?,?,?)",
                [
                    (speaker_id, audio_sha, vector.astype("<f8").tobytes())
                    for audio_sha, vector in prepared
                ],
            )
            revision = self._advance()
        return {
            "operation": "enrolled",
            "speaker_id": speaker_id,
            "samples": count + len(prepared),
            "state": "ready"
            if count + len(prepared) >= self.policy.minimum_enrollment_samples
            else "collecting",
            "revision": revision,
        }

    def _snapshot(self):
        # BEGIN is necessary: separate SELECTs must not see different commits.
        with self.db:
            self.db.execute("BEGIN")
            revision = self._revision()
            if self.db.execute("PRAGMA foreign_key_check").fetchone():
                raise SpeakerIdentityError("corrupt enrollment references")
            rows = self.db.execute(
                "SELECT p.id,p.label,s.audio_sha,s.embedding FROM speakers p LEFT JOIN samples s ON s.speaker_id=p.id ORDER BY p.id,s.audio_sha LIMIT ?",
                (MAX_SPEAKERS * MAX_SAMPLES + 1,),
            ).fetchall()
        groups = {}
        for sid, label, audio_sha, blob in rows:
            valid_id(sid)
            if (
                not isinstance(label, str)
                or not 1 <= len(label) <= 128
                or any(ord(c) < 32 for c in label)
            ):
                raise SpeakerIdentityError("corrupt enrollment label")
            sha_string(audio_sha)
            if not isinstance(blob, bytes) or len(blob) != DIMENSIONS * 8:
                raise SpeakerIdentityError("corrupt enrollment embedding")
            raw = np.frombuffer(blob, dtype="<f8")
            if not np.isfinite(raw).all() or abs(float(np.linalg.norm(raw)) - 1) > 1e-6:
                raise SpeakerIdentityError("corrupt enrollment unit vector")
            item = groups.setdefault(sid, {"label": label, "vectors": []})
            item["vectors"].append(raw)
        if len(groups) > MAX_SPEAKERS or any(
            not 1 <= len(g["vectors"]) <= MAX_SAMPLES for g in groups.values()
        ):
            raise SpeakerIdentityError("corrupt enrollment cardinality")
        return revision, groups

    def list(self):
        revision, groups = self._snapshot()
        return {
            "revision": revision,
            "speakers": [
                {"speaker_id": sid, "label": g["label"], "samples": len(g["vectors"])}
                for sid, g in groups.items()
            ],
        }

    def identify(self, embedding, *, embedding_binding: str) -> Decision:
        if embedding_binding != self.policy.embedding_binding:
            raise SpeakerIdentityError("embedding model/frontend binding differs")
        query = unit(embedding)
        revision, groups = self._snapshot()
        common = {
            "enrollment_revision": revision,
            "policy_sha256": self.policy.fingerprint,
        }
        if not groups:
            return Decision(
                "unknown", None, None, None, None, "no_enrollments", **common
            )
        scores = sorted(
            (
                (
                    float(np.clip(query @ unit(np.mean(g["vectors"], axis=0)), -1, 1)),
                    sid,
                )
                for sid, g in groups.items()
            ),
            key=lambda row: (-row[0], row[1]),
        )
        score, sid = scores[0]
        margin = score - scores[1][0] if len(scores) > 1 else None
        if len(groups[sid]["vectors"]) < self.policy.minimum_enrollment_samples:
            return Decision(
                "unknown",
                None,
                None,
                score,
                margin,
                "insufficient_enrollment",
                **common,
            )
        if score < self.policy.threshold:
            return Decision(
                "unknown",
                None,
                None,
                score,
                margin,
                "below_acceptance_threshold",
                **common,
            )
        # A tied nearest neighbour is never an identity, even with margin=0.
        if margin is not None and (margin <= 0 or margin < self.policy.minimum_margin):
            return Decision(
                "ambiguous",
                None,
                None,
                score,
                margin,
                "insufficient_separation",
                **common,
            )
        return Decision(
            "known", sid, groups[sid]["label"], score, margin, "accepted", **common
        )

    def remove(self, speaker_id: str):
        valid_id(speaker_id)
        with self.db:
            self.db.execute("BEGIN IMMEDIATE")
            revision = self._revision()
            removed = self.db.execute(
                "DELETE FROM speakers WHERE id=?", (speaker_id,)
            ).rowcount
            if removed:
                revision = self._advance()
        return {
            "operation": "removed",
            "speaker_id": speaker_id,
            "removed": bool(removed),
            "revision": revision,
        }

    def reset(self):
        with self.db:
            self.db.execute("BEGIN IMMEDIATE")
            self._revision()
            count = self.db.execute("SELECT COUNT(*) FROM speakers").fetchone()[0]
            self.db.execute("DELETE FROM speakers")
            revision = self._advance()
        return {"operation": "reset", "removed_speakers": count, "revision": revision}
